tests/testthat/test-epi_slide.R

```diff
@@ -4,31 +4,55 @@ edf = dplyr::bind_rows(
 
 ## --- These cases generate the error: ---
+test_that("`after` must be defined as a non-zero integer if `before` is missing", {
+  expect_error(epi_slide(edf, f, after = 0L, ref_time_values=as.Date("2020-01-01")),
+               "`before` cannot be missing when `after` is set to 0.")
+})
+
+test_that("Warn user against having a blank `before`",{
+  expect_warning(epi_slide(edf, f, after = 1L, ref_time_values=as.Date("2020-01-01")+1L),
+                 regexp="`before` missing but `after` nonzero; `before` has been set to 0.")
+})
+
+test_that("Both `before` and `after` must be nonnegative integers",{
+  expect_error(epi_slide(edf, f, before = -1L, ref_time_values=as.Date("2020-01-01")+2L),
+               "`before` and `after` must be at least 0.")
+  expect_error(epi_slide(edf, f, before = 2L, after = -1L, ref_time_values=as.Date("2020-01-01")+2L),
+               "`before` and `after` must be at least 0.")
+  expect_error(epi_slide(edf, f, before = 0.5, ref_time_values=as.Date("2020-01-01")+2L),
+               "`before` and `after` must be integers.")
+  expect_error(epi_slide(edf, f, before = 1, after = 0.5, ref_time_values=as.Date("2020-01-01")+2L),
+               "`before` and `after` must be integers.")
+  # The before and after values can be numerics that are integerish
+  expect_error(epi_slide(edf, f, before = 1, after = 1, ref_time_values=as.Date("2020-01-01")+2L),NA)
+})
+
 test_that("`ref_time_values` + `align` that result in no slide data, generate the error", {
-  expect_error(edf %>% group_by(geo_value) %>% epi_slide(f, n=3L, ref_time_values=as.Date("2020-01-01")), 
+  expect_error(epi_slide(edf, f, before=2L, ref_time_values=as.Date("2020-01-01")), 
                "starting and/or stopping times for sliding are out of bounds") # before the first, no data in the slide windows
-  expect_error(edf %>% group_by(geo_value) %>% epi_slide(f, n=3L, ref_time_values=as.Date("2020-01-01")+207L), 
+  expect_error(epi_slide(edf, f, before=2L, ref_time_values=as.Date("2020-01-01")+207L), 
                "starting and/or stopping times for sliding are out of bounds") # beyond the last, no data in window
 })
 
 test_that("`ref_time_values` + `align` that have some slide data, but generate the error due to ref. time being out of time range", {
-  expect_error(edf %>% group_by(geo_value) %>% epi_slide(f, n=3L, ref_time_values=as.Date("2020-01-01"), align="left"), 
+  expect_error(epi_slide(edf, f, before=0L, after=2L, ref_time_values=as.Date("2020-01-01")), 
                "starting and/or stopping times for sliding are out of bounds") # before the first, but we'd expect there to be data in the window
-  expect_error(edf %>% group_by(geo_value) %>% epi_slide(f, n=3L, ref_time_values=as.Date("2020-01-01")+201L), 
+  expect_error(epi_slide(edf, f, before=2L, ref_time_values=as.Date("2020-01-01")+201L), 
                "starting and/or stopping times for sliding are out of bounds") # beyond the last, but still with data in window
```

Comment:
I think at some point we want to change this behavior, right?  Do you remember if this is already filed in an issue?

Reply:
We would want errors if there is no data in the window, but if there is partially data in the window, I assume we'd either want to uniformly allow it or uniformly reject it, but currently we aren't consistent (we allow partial data in window it for a trailing window for ref time values near min(time_value) in the edf, but don't allow it for manually-input ref time values slightly beyond the max(time_value)).  In any case, something for another issue.

Reply:
What behaviour are you talking about? I don't quite understand what you mean here....

Reply:
Side note: saw an unrelated(?) pre-existing bug: some instances to `unique(x$time_value)` likely need to be sorted to be robust.

Current behavior:
- check for out-of-range start/stop time values only triggers if they are ALL out of range.
- a user-provided ref_time_values is silently filtered to down to those in unique(x$time_value)`.

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(epiprocess)
#> 
#> Attaching package: 'epiprocess'
#> The following object is masked from 'package:stats':
#> 
#>     filter
edf = tibble(geo_value="g1", time_value = 1:50, value=1:50) %>% as_epi_df()
# Silently excludes 51 (which would have partial data in window if we allowed
# it), but includes 1 (which has partial data):
edf %>% epi_slide(~ toString(.x$value), before=2L, ref_time_values=c(1,50,51))
#> An `epi_df` object, 2 x 4 with metadata:
#> * geo_type  = custom
#> * time_type = custom
#> * as_of     = 2022-08-15 15:30:44
#> 
#> # A tibble: 2 × 4
#>   geo_value time_value value slide_value
#> * <chr>          <int> <int> <chr>      
#> 1 g1                 1     1 1          
#> 2 g1                50    50 48, 49, 50
# Triggers error:
edf %>% epi_slide(~ toString(.x$value), before=2L, ref_time_values=51)
#> Error in `Abort()` at epiprocess/R/slide.R:189:4:
#> ! The starting and/or stopping times for sliding are out of bounds with
#> respect to the range of times in your data. Check your settings for
#> ref_time_values and align (and before, if specified).
```

<sup>Created on 2022-08-15 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

I'm hoping for something more consistent.  Could you please make sure that the two things above (sorting the unique times when appropriate, and making the above more consistent) are filed as issues?

Reply:
I added issues to both.